From 4e0bf723e6c4c5beac6f3ccc74905fc9f14dd0eb Mon Sep 17 00:00:00 2001
From: Jonas Gorski <jonas.gorski@bisdn.de>
Date: Tue, 15 Nov 2022 14:53:49 +0100
Subject: [PATCH 7/7] delta-ag5648: update i2c_driver with 6.1 compatibility

i2c_remove changes its return type from int to void, so we need to
provide different versions based on kernel version.

Add variants using the following semantic patch:

// <smpl>
@r@
identifier DRIVER;
identifier remove_fn;
fresh identifier remove_fn_wrap = remove_fn ## "_6_1";
position p;
@@
struct i2c_driver DRIVER@p = {
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6,1,0)
+     .remove = remove_fn_wrap,
+#else
      .remove = remove_fn,
+#endif
};

@i depends on r@
@@
#include <linux/version.h>
@depends on r&&!i@
@@
#include <...>
+#include <linux/version.h>
@depends on r@
identifier r.remove_fn;
identifier r.remove_fn_wrap;
@@
int remove_fn(...) {...}
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6,1,0)
+static void remove_fn_wrap(struct i2c_Client *client)
+{
+     remove_fn(client);
+}
+#endif
// </smpl>

applied with

  spatch --smpl-spacing --sp-file 6.1-i2c_remove_sig.cocci --in-place --dir packages/

Result taken as is, without any code style fixes.

Manually added missing prototype for new function.

Signed-off-by: Jonas Gorski <jonas.gorski@bisdn.de>
---
 .../x86-64/ag5648/modules/builds/dni_ag5648_psu.c  | 11 +++++++++++
 .../x86-64/ag5648/modules/builds/dni_ag5648_sfp.c  | 11 +++++++++++
 .../x86-64/ag5648/modules/builds/dni_emc2305.c     | 14 ++++++++++++++
 .../delta/x86-64/modules/builds/i2c_cpld.c         | 10 ++++++++++
 4 files changed, 46 insertions(+)

diff --git a/packages/platforms/delta/x86-64/ag5648/modules/builds/dni_ag5648_psu.c b/packages/platforms/delta/x86-64/ag5648/modules/builds/dni_ag5648_psu.c
index 1a25071a090a..d26e9ebbddb8 100755
--- a/packages/platforms/delta/x86-64/ag5648/modules/builds/dni_ag5648_psu.c
+++ b/packages/platforms/delta/x86-64/ag5648/modules/builds/dni_ag5648_psu.c
@@ -34,6 +34,7 @@
 #include <linux/hwmon.h>
 #include <linux/hwmon-sysfs.h>
 #include <linux/err.h>
+#include <linux/version.h>
 
 #define MAX_FAN_DUTY_CYCLE 100
 
@@ -456,6 +457,12 @@ static int dps_800ab_16_d_remove(struct i2c_client *client)
 	
 	return 0;
 }
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6,1,0)
+static void dps_800ab_16_d_remove_6_1(struct i2c_client *client)
+{
+	dps_800ab_16_d_remove(client);
+}
+#endif
 
 enum id_name {
 	dni_ag5648_psu,
@@ -476,7 +483,11 @@ static struct i2c_driver dps_800ab_16_d_driver = {
                 .name   = "dps_800ab_16_d",
         },
         .probe          = dps_800ab_16_d_probe,
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6,1,0)
+        .remove = dps_800ab_16_d_remove_6_1,
+#else
         .remove         = dps_800ab_16_d_remove,
+#endif
         .id_table       = dps_800ab_16_d_id,
         .address_list   = normal_i2c,
 };
diff --git a/packages/platforms/delta/x86-64/ag5648/modules/builds/dni_ag5648_sfp.c b/packages/platforms/delta/x86-64/ag5648/modules/builds/dni_ag5648_sfp.c
index f3f06153865e..6dc38ec6d1d5 100755
--- a/packages/platforms/delta/x86-64/ag5648/modules/builds/dni_ag5648_sfp.c
+++ b/packages/platforms/delta/x86-64/ag5648/modules/builds/dni_ag5648_sfp.c
@@ -32,6 +32,7 @@
 #include <linux/hwmon.h>
 #include <linux/hwmon-sysfs.h>
 #include <linux/err.h>
+#include <linux/version.h>
 
 #define I2C_BUS_2 2
 #define MASTER_CPLD 0x35
@@ -656,6 +657,12 @@ static int ag5648_sfp_remove(struct i2c_client *client)
 	kfree(data);
 	return 0;
 }
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6,1,0)
+static void ag5648_sfp_remove_6_1(struct i2c_client *client)
+{
+	ag5648_sfp_remove(client);
+}
+#endif
 
 enum id_name 
 {
@@ -675,7 +682,11 @@ static struct i2c_driver ag5648_sfp_driver = {
 		.name	= "dni_ag5648_sfp",
 	},
 	.probe		= ag5648_sfp_probe,
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6,1,0)
+	.remove = ag5648_sfp_remove_6_1,
+#else
 	.remove		= ag5648_sfp_remove,
+#endif
 	.id_table	= ag5648_sfp_id,
 	.address_list	= normal_i2c,
 };
diff --git a/packages/platforms/delta/x86-64/ag5648/modules/builds/dni_emc2305.c b/packages/platforms/delta/x86-64/ag5648/modules/builds/dni_emc2305.c
index ba63bdc8e50a..edd46a60421f 100755
--- a/packages/platforms/delta/x86-64/ag5648/modules/builds/dni_emc2305.c
+++ b/packages/platforms/delta/x86-64/ag5648/modules/builds/dni_emc2305.c
@@ -22,6 +22,7 @@
 #include <linux/hwmon.h>
 #include <linux/hwmon-sysfs.h>
 #include <linux/err.h>
+#include <linux/version.h>
 
 extern int i2c_cpld_read(int bus, unsigned short cpld_addr, u8 reg);
 extern int i2c_cpld_write(int bus, unsigned short cpld_addr, u8 reg, u8 value);
@@ -73,6 +74,9 @@ static int emc2305_probe(struct i2c_client *client,
 static int emc2305_detect(struct i2c_client *client,
                           struct i2c_board_info *info);
 static int emc2305_remove(struct i2c_client *client);
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6,1,0)
+static void emc2305_remove_6_1(struct i2c_client *client);
+#endif
 
 static const struct i2c_device_id emc2305_id[] =
 {
@@ -88,7 +92,11 @@ static struct i2c_driver emc2305_driver =
     .name = "emc2305",
   },
   .probe    = emc2305_probe,
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6,1,0)
+  .remove = emc2305_remove_6_1,
+#else
   .remove   = emc2305_remove,
+#endif
   .id_table = emc2305_id,
   .detect   = emc2305_detect,
   .address_list = normal_i2c,
@@ -376,6 +384,12 @@ static int emc2305_remove(struct i2c_client *client)
   sysfs_remove_group(&client->dev.kobj, &data->attrs);
   return 0;
 }
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6,1,0)
+static void emc2305_remove_6_1(struct i2c_client *client)
+{
+  emc2305_remove(client);
+}
+#endif
 
 module_i2c_driver(emc2305_driver);
 
diff --git a/packages/platforms/delta/x86-64/modules/builds/i2c_cpld.c b/packages/platforms/delta/x86-64/modules/builds/i2c_cpld.c
index bf368a781dc9..159f05a4f926 100755
--- a/packages/platforms/delta/x86-64/modules/builds/i2c_cpld.c
+++ b/packages/platforms/delta/x86-64/modules/builds/i2c_cpld.c
@@ -417,6 +417,12 @@ static int i2c_cpld_mux_remove(struct i2c_client *client)
 
   return 0;
 }
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6,1,0)
+static void i2c_cpld_mux_remove_6_1(struct i2c_client *client)
+{
+  i2c_cpld_mux_remove(client);
+}
+#endif
 
 int i2c_cpld_read(int bus, unsigned short cpld_addr, u8 reg)
 {
@@ -477,7 +483,11 @@ static struct i2c_driver i2c_cpld_mux_driver =
     .owner  = THIS_MODULE,
   },
   .probe    = i2c_cpld_mux_probe,
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6,1,0)
+  .remove = i2c_cpld_mux_remove_6_1,
+#else
   .remove   = i2c_cpld_mux_remove,
+#endif
   .id_table = i2c_cpld_mux_id,
 };
 
-- 
2.38.1

